### PR TITLE
Add Hindi (hi-IN) support for TTS

### PR DESCRIPTION
**Pull request by quapham**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add Hindi (hi-IN) TTS support with char tokenizer 

**Collection**: [Note which collection this PR will affect]
TTS, common
# Changelog
- nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py
Added Hindi grapheme character set (`hi-IN`) with Devanagari script support

- nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py
"HindiCharsTokenizer" class for Hindi grapheme tokenization
Handles Unicode decomposition correctly (e.g., ड़ = ड + ़ nukta)

# Usage

- You can potentially add a usage example below

```python
# Code tests for Hindi character tokenizer
from nemo.collections.common.tokenizers.text_to_speech.tts_tokenizers import HindiCharsTokenizer
tokenizer = HindiCharsTokenizer(
    punct=True,
    apostrophe=True,
    pad_with_space=False
)
text="अंगड़ाई"
ids = tokenizer.encode(text)
decoded = tokenizer.decode(ids)

print(f"IDs:   {ids}")
print(f"Decoded: {decoded}")

#IDS: [74, 138, 90, 100, 141, 124, 77]
#Decoded:अंगड़ाई
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by XuesongYang** [comment]
could you pls add unit tests to verify if they worked as expected? examples are,
1. https://github.com/NVIDIA-NeMo/NeMo/blob/main/tests/collections/tts/g2p/test_modules.py
2. https://github.com/NVIDIA-NeMo/NeMo/blob/main/tests/collections/common/tokenizers/text_to_speech/test_tts_tokenizers.py

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py [hunk 1]
```diff
@@ -147,15 +163,22 @@
...
+        'w', 'k', 'ɾ', 's', 't', 'ʃ', 'r', 'h', 'n', 'nʲ',
         'ɲ', 'ç', 'b', 'm', 'j', 'ɸ', 'z', 'p', 'd', 'N',
         'ʒ', 'ŋ', 'g', 'f', 'ʔ', 'y', 'ɟ', 'v', 'ɥ', 'ɰ',
         'ɰᵝ', 'ɣ', 'ʄ', 'ʑ', 'c', 'ɕ', 'ɠ', 'x', 'l', 'β',
         'ð', 'ø', 'ʁ', 'ts', 'tʃ', 'dʒ', 'y', 'dʑ', 't͡s',
-        'ɑ̃', 'ĩ', 'ũ', 'ẽ', 'õ', 'ɑ̃', 'ĩ', 'ũ', 'w̃',  
-        'ẽ', 'õ', 'hʲ', 'ɪ', 'ː', 'o̞', 'e̞', 
+        'ɑ̃', 'ĩ', 'ũ', 'ẽ', 'õ', 'ɑ̃', 'ĩ', 'ũ', 'w̃',
+        'ẽ', 'õ', 'hʲ', 'ɪ', 'ː', 'o̞', 'e̞',
+    ),
```
[the remark this answers is not in the document]
Comment (XuesongYang) [reply nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:175]: @quapham could you pls add this notes if Copilot is right?

File: nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py [hunk 2]
```diff
@@ -380,6 +380,81 @@ def __init__(
...
+            # Add a punctuation that has a single char.
+            elif (c in self.PUNCT_LIST) and self.punct:
+                cs.append(c)
+
+            elif c != space:
+                logging.warning(f"Text: [{text}] contains unknown char: [{c}]. Symbol will be skipped.")
+
+        # Remove trailing spaces
+        if cs:
+            while cs[-1] == space:
```
[the remark this answers is not in the document]
Comment (XuesongYang) [reply nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py:449]: `if cs` alreay covers the test of empty list.

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py [hunk 3]
```diff
@@ -147,15 +163,22 @@
...
+    "hi-IN": (
+        '.', 'a', 'b', 'c', 'd', 'e', 'f', 'h', 'i', 'j',
+        'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
+        'u', 'w', 'x', 'z', 'ŋ', 'ɔ', 'ɖ', 'ə', 'ɛ', 'ɟ',
+        'ɡ', 'ɣ', 'ɪ', 'ɭ', 'ɲ', 'ɳ', 'ɾ', 'ʂ', 'ʃ', 'ʈ',
+        'ʊ', 'ʋ', 'ʌ', 'ʰ', 'ː', '̃', '̩', 'χ',
     ),
```
Comment (XuesongYang) [inline nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:185]: "hi-IN" is defined in `IPA_CHARACTER_SETS`, so I guess you may have implemented phoneme tokenizer as well. Could you pls also add below items if any?

1. Hindi's IPA dictionary;
2. if no phoneme tokenizer is defined, pls ensure `IpaG2p` and `IPATokenizer` correctly support Hindi.
3. add unit test following `test_ipa_tokenizer_es_es`.

Comment (quapham) [reply nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:185]: Yes, it will have a dictionary. Currently, it works in our tests, so I will add this as well.

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (quapham) [reply nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:175]: I extracted all IPA symbols directly from the hindi IPA dictionary. The "." is part of the IPA, so it is correct. We can keep this note.

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py (hunk 3 above)
Comment (XuesongYang) [reply nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:185]: nice. pls add those changes.

**Comment by subhankar-ghosh** [comment]
> Let's pause to merge this PR for now. We need to add hi-IN dict and unit tests.
> 
> @quapham needs more time running additional tests to make sure the checkpoint hindi IPA is good.

@quapham Did you get a chance to run the additional tests?

**Comment by subhankar-ghosh** [comment]
This PR is just for Char tokenizer correct? and NOT phoneme tokenizer. @quapham You have the phoneme dict but it is not being used anywhere in this PR. Is that the correct understanding?

**Comment by quapham** [comment]
Thank @XuesongYang . I will create a new PR for the IPA G2P phoneme.
@subhankar-ghosh, you are correct, this PR currently only supports the Hindi chartokenizer.

**Comment by chtruong814** [comment]
The test failure is likely temporary network issue.  It's a 404 from huggingface on a test that has worked before.  Will go ahead and merge this.
